tests/components/dlna_dmr/test_media_player.py

```diff
@@ -760,6 +764,89 @@ async def test_play_media_metadata(
     )
 
 
+async def test_play_media_local_source(
+    hass: HomeAssistant, dmr_device_mock: Mock, mock_entity_id: str
+) -> None:
+    """Test play_media with a media_id from a local media_source."""
+    # Based on roku's test_services_play_media_local_source and cast's
+    # test_entity_browse_media
+    await async_setup_component(hass, MS_DOMAIN, {MS_DOMAIN: {}})
+    await hass.async_block_till_done()
+
+    await hass.services.async_call(
+        MP_DOMAIN,
+        mp_const.SERVICE_PLAY_MEDIA,
+        {
+            ATTR_ENTITY_ID: mock_entity_id,
+            mp_const.ATTR_MEDIA_CONTENT_TYPE: "video/mp4",
+            mp_const.ATTR_MEDIA_CONTENT_ID: "media-source://media_source/local/Epic Sax Guy 10 Hours.mp4",
+        },
+        blocking=True,
+    )
+
+    assert dmr_device_mock.construct_play_media_metadata.await_count == 1
+    assert (
+        "/media/local/Epic%20Sax%20Guy%2010%20Hours.mp4?authSig="
+        in dmr_device_mock.construct_play_media_metadata.call_args.kwargs["media_url"]
+    )
+    assert dmr_device_mock.async_set_transport_uri.await_count == 1
+    assert dmr_device_mock.async_play.await_count == 1
+    call_args = dmr_device_mock.async_set_transport_uri.call_args.args
+    assert "/media/local/Epic%20Sax%20Guy%2010%20Hours.mp4?authSig=" in call_args[0]
+
+
+async def test_play_media_didl_metadata(
+    hass: HomeAssistant, dmr_device_mock: Mock, mock_entity_id: str
+) -> None:
+    """Test play_media passes available DIDL-Lite metadata to the DMR."""
+
+    @dataclass
+    class DidlPlayMedia(PlayMedia):
+        """Playable media with DIDL metadata."""
+
+        didl_metadata: didl_lite.DidlObject
+
+    didl_metadata = didl_lite.VideoItem(
+        id="120$22$33",
+        restricted="false",
+        title="Epic Sax Guy 10 Hours",
+        res=[
+            didl_lite.Resource(uri="unused-URI", protocol_info="http-get:*:video/mp4:")
+        ],
+    )
+
+    play_media = DidlPlayMedia(
+        url="/media/local/Epic Sax Guy 10 Hours.mp4",
+        mime_type="video/mp4",
+        didl_metadata=didl_metadata,
+    )
+
+    await async_setup_component(hass, MS_DOMAIN, {MS_DOMAIN: {}})
+    await hass.async_block_till_done()
+    local_source = hass.data[MS_DOMAIN][MS_DOMAIN]
```

Comment:
We don't want to access hass.data in tests. Can we avoid this, eg by changing how we patch? Maybe patch `media_source.async_resolve_media`?

Reply:
Follow-up PR: https://github.com/home-assistant/core/pull/66939